### ISSUE TITLE
tracing with async php runtimes

### ISSUE THREAD

**Issue by brettmc** [open-telemetry/opentelemetry-php#671]
**Describe your environment**
Using an async (not-shared-nothing) PHP runtime (eg roadrunner, swoole, react) where a long-lived application is used to service multiple requests.
Since we use a single shared `ContextStorage`, a call to `$span->activate()` will set the active span in that single context storage. Then, any other requests executed in the same process/worker will be parented to that span, and attempts to set an active span will clobber any other request's active span.

**Steps to reproduce**
Any simple otel-instrumented PHP application which creates a span, when run in parallel in the one runtime.

**What is the expected behavior?**
Traces and spans should be correctly nested, as if they were run in an isolated process in a shared-nothing runtime (apache, fpm)

**What is the actual behavior?**
Generated traces and spans are randomly parented to each other in the order they were executed.

**Additional context**
Our usage of `Context::getCurrent()` looks to be the core of the problem, since it always operates on a single, static storage.

I've spent some time looking at this, and I think the path forward is to:
- allow the creation of individual storage (eg one per request/operation)
- default to using a static storage (as we do now, which is fine for apache/fpm)
- update classes which use `Context::getCurrent()` to instead use `$storage->current()`, and allow those classes/methods to accept storage as a class variable (or static method parameter), so that "get the current context" becomes "get the current context from this storage"

**Comment by tidal** [comment]
That sounds like a good idea.
I would like if we also can make use of the strength of roadrunner, swoole, react, etc.
So what about we create adapter for the different runtimes?
For the runtime detectors, we just need to implement them once for the attributes,
and we could reuse the detectors.
Also we may introduce a `ContextProvider` so we have a consistent pattern across all the components.

**Comment by brettmc** [comment]
It's an interesting idea to be able to detect a runtime, and ContextProvider might be a good abstraction to hide storage complexities, but I do think that `ContextStorage` is the key.
I've got some changes in a branch which work (can have multiple traces being created at the same time), but it's not the most elegant solution - I will submit as a draft PR to seek feedback. The idea is that a static default storage is used if you don't do anything (as happens now, so the API is unchanged for apache/fpm/etc).
If you do require concurrency, then you are responsible for managing that and keeping track of each request's context/storage/tracer, for example by:
- creating per-request storage, keeping track of it (eg in a request attribute), passing it to `spanBuilder` (which is what I have now); or perhaps
- creating per-request tracer with individual storage attached (not to spec?), keep track of it as above

I can't think of a way which avoids requiring each request to keep track of its context/storage

**Comment by Grunet** [comment]
Fwiw from what I understand of otel-js so far (for node and browsers) they seem to require consumers to be pretty explicit about creating/passing context around for connecting parent/child spans (at least from what I can tell from their examples at least, like this one - https://github.com/open-telemetry/opentelemetry-js/blob/main/examples/basic-tracer-node/index.js#L48 - and using their api a bit)

But yeah for traditional PHP runtimes that seems like it could get annoying pretty fast...

**Comment by Nevay** [comment]
Execution switches should preferably be handled within the global `ContextStorage` (as already done for `Fiber`s; these kinds of execution switches were the reason why I created the context storage abstraction in the first place).

A basic implementation for Swoole can be found [here](https://github.com/Nevay/opentelemetry-php/blob/e8fa4fd09a3971ab6ebb1fe224e34d356d998c7c/src/Context/Swoole/SwooleContextHandler.php#L24-L36); incompatible with fibers. Swoole doesn't seem to provide a way to hook into coroutine creation which complicates supporting Fibers and Swoole at the same time (I am not familiar with Swoole, please correct me if I am wrong).
I can create a PR if I find time to deal with the remaining issues (adding tests similar to the tests for fibers and making phan/psalm/phpstan happy (PhanUndeclaredClassReference & co.)) or if someone is willing to continue the work.

**Comment by brettmc** [comment]
@Nevay thanks for the update. I'm not clear on whether what you're working on is a complete replacement for #674 ... I did consider using the forking concept, but I did not get very far (I'm wary of my approach's requirement to  create infinite ContextStorage instances, so I certainly see the benefits if it can work). If you can find a few minutes to catch me on CNCF slack and get me up to speed on what you're doing, I'll be happy to assist.
